provider/ultradns/ultradns.go

```diff
@@ -0,0 +1,502 @@
+			return nil, fmt.Errorf("please set proper act on probe value, the values can be either true or false")
+		}
+		sbPoolActOnProbes, _ = strconv.ParseBool(actOnProbeValue)
+	}
+
+	poolValue, ok := os.LookupEnv("ULTRADNS_POOL_TYPE")
+	if ok {
+		if (poolValue != "sbpool") && (poolValue != "rdpool") {
+			return nil, fmt.Errorf(" please set proper ULTRADNS_POOL_TYPE, supported types are sbpool or rdpool")
+		}
+		ultradnsPoolType = poolValue
+	}
+
+	client, err := udnssdk.NewClient(Username, string(Password), BaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("connection cannot be established")
+	}
+
+	provider := &UltraDNSProvider{
+		client:       *client,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+		AccountName:  AccountName,
+	}
+
+	return provider, nil
+}
+
+// Zones returns list of hosted zones
+func (p *UltraDNSProvider) Zones(ctx context.Context) ([]udnssdk.Zone, error) {
+	zoneKey := &udnssdk.ZoneKey{}
+	var err error
+
+	if p.domainFilter.IsConfigured() {
+		zonesAppender := []udnssdk.Zone{}
+		for _, zone := range p.domainFilter.Filters {
+			zoneKey.Zone = zone
+			zoneKey.AccountName = p.AccountName
+			zones, err := p.fetchZones(ctx, zoneKey)
+
+			if err != nil {
+				return nil, err
+			}
+
+			zonesAppender = append(zonesAppender, zones...)
+		}
+		return zonesAppender, nil
+	}
+	zoneKey.AccountName = p.AccountName
+	zones, err := p.fetchZones(ctx, zoneKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return zones, nil
+}
+
+func (p *UltraDNSProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	zones, err := p.Zones(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		log.Infof("zones : %v", zone)
+		rrsetType := ""
+		ownerName := ""
+		rrsetKey := udnssdk.RRSetKey{
+			Zone: zone.Properties.Name,
+			Type: rrsetType,
+			Name: ownerName,
+		}
+
+		if zone.Properties.ResourceRecordCount != 0 {
+			records, err := p.fetchRecords(ctx, rrsetKey)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, r := range records {
+				recordTypeArray := strings.Fields(r.RRType)
+				if provider.SupportedRecordType(recordTypeArray[0]) {
+					log.Infof("owner name %s", r.OwnerName)
+					name := r.OwnerName
+
+					// root name is identified by the empty string and should be
+					// translated to zone name for the endpoint entry.
+					if r.OwnerName == "" {
+						name = zone.Properties.Name
+					}
+
+					endPointTTL := endpoint.NewEndpointWithTTL(name, recordTypeArray[0], endpoint.TTL(r.TTL), r.RData...)
+					endpoints = append(endpoints, endPointTTL)
+				}
+			}
+		}
+	}
+	log.Infof("endpoints %v", endpoints)
+	return endpoints, nil
+}
+
+func (p *UltraDNSProvider) fetchRecords(ctx context.Context, k udnssdk.RRSetKey) ([]udnssdk.RRSet, error) {
+	// TODO: Sane Configuration for timeouts / retries
+	maxerrs := 5
+	waittime := 5 * time.Second
+
+	rrsets := []udnssdk.RRSet{}
+	errcnt := 0
+	offset := 0
+	limit := 1000
+
+	for {
```

Comment:
why you want to have this? we have a reconciliation loop which always goes into each provider

Reply:
The UltraDNS API  by default get 1000 as limit, so we need to paginate through all of the results so we have added this for loop